SVA/lib/mmu.c

```diff
@@ -1379,6 +1312,160 @@ unmapSecurePage (struct SVAThread * threadp, unsigned char * v) {
   return;
 }
 
+/*   Function: ghostmemCOW()
+ *   
+ *   Description: 
+ *   Copy the parent's page table of ghost memory to the child. 
+ *   Write protect these page table entries for both the parent and the child.
+ *
+ *   Inputs:
+ *   oldThread - the SVAThread variable of the parent process
+ *   newThread - the SVAThread variable of the child process   
+ */
+void
+ghostmemCOW(struct SVAThread* oldThread, struct SVAThread* newThread)
+{
+    
+    uintptr_t vaddr_start, vaddr_end, size;
+    
+    vaddr_start = (uintptr_t) SECMEMSTART;
+    size = oldThread->secmemSize;
+    vaddr_end = vaddr_start + size;    
+
+   /*
+    * Get the PML4E of the new process's page table.  If there isn't one in the
+    * table, add one.
+    */
+    sva_integer_state_t integerState = newThread->integerState;
+    pml4e_t * pml4e =  (pml4e_t *) getVirtual(integerState.cr3 + secmemOffset);
```

Comment:
Is this where you're using the CR3 field of the Integer State?  If so, why not use the secmemPML4e in the SVAThread to find the page table entries that map ghost memory?

Reply:
ghostmemCOW() is called by sva_init_stack() at fork, where the new thread (child)'s SVAThread->secmemPML4e is set to be equal to the old thread's. The new thread's secmemPML4e is not set until in sva_ghost_fault() or ghostMalloc() when the ghost memory is allocated at the first time. Therefore I am not able to know the new child's pml4 at fork without passing it from the OS.